File: LeetCode/1288-remove-covered-intervals.c
#include "_leetcode_common.c"

int sorter(const void *r1, const void *r2) {
  int *_r1 = *(int **) r1, *_r2 = *(int **) r2;
  return _r1[0] != _r2[0] ? _r1[0] - _r2[0] : _r2[1] - _r1[1];
}

/**
 * Determines whether a covers b.
 */
static inline bool is_covered(int *r1, int *r2) {
  return r1[0] <= r2[0] && r2[1] <= r1[1];
}

int removeCoveredIntervals(int **intervals, int intervalsSize, int *intervalsColSize) {
  if (intervalsSize == 0)
    return 0;

  qsort(intervals, intervalsSize, 2 * sizeof(int), sorter);

  int res = intervalsSize;
  int *prev = intervals[0];
  for (int i = 1; i < intervalsSize; ++i) {
    if (is_covered(prev, intervals[i])) {
      --res;
    } else {
      prev = intervals[i];
    }
  }

  return res;
}

int main(int argc, char *argv[]) {
  return 0;
}

File: LeetCode/1288-remove-covered-intervals.py
from functools import cmp_to_key
from typing import List


class Solution:
    def removeCoveredIntervals(self, intervals: List[List[int]]) -> int:
        def is_covered(a: List[int], b: List[int]) -> bool:
            """Determines whether a covers b."""
            return a[0] <= b[0] and b[1] <= a[1]

        if not intervals:
            return 0

        res = len(intervals)
        intervals.sort(
            key=cmp_to_key(lambda a, b: a[0] - b[0] if a[0] != b[0] else b[1] - a[1])
        )

        prev = intervals[0]
        for i in range(1, res):
            if is_covered(prev, intervals[i]):
                res -= 1
            else:
                prev = intervals[i]

        return res


s = Solution()
res = s.removeCoveredIntervals([[1, 5], [1, 6], [5, 7]])
print(f"{res = }")
